### DISCUSSION TITLE
Pico-PIO-USB as an USB man-in-the-middle?

### DISCUSSION THREAD

**Discussion by eliocor** [sekigon-gonnoc/Pico-PIO-USB#30] [Ideas]
Is there any way to program a Pico board as to act as a MITM (man-in-the-middle) between an USB device and its host?
It will be useful as a debugging tool to monitor/inject/modify data from/to the device/host as was doing [USBProxy](https://github.com/MatrixCat3D/USBProxy) on the beaglebone: see also https://www.youtube.com/watch?v=5JnAeakUBnU

**Comment by rppicomidi** [comment]
@eliocor It is not a general purpose MITM USBProxy, but I did implement something that modifies data between a MIDI host and MIDI device. See [pico-usb-midi-filter](https://github.com/rppicomidi/pico-usb-midi-filter). I implemented hooks in the USB stack so that the USB device descriptor the Pico board presents to the PC is the same descriptor that the Pico-PIO-USB host reads from the external MIDI device.

**Comment by ax16** [comment]
I hope that someday, someone will make a fork dedicated to using pico as MITM USB proxy. I checked out [pico-usb-midi-filter](https://github.com/rppicomidi/pico-usb-midi-filter) but i am unfortunately not experienced enough in C to create it myself.

**Comment by Gao-Ruilin** [comment]
Hello, I'm trying to do this, and it's going well. I'll let you know when it's ready to use.
